src/CommandLineParser.ts

```diff
@@ -0,0 +1,169 @@
+        type: 'boolean'
+    },
+    'h': {
+        alias: 'help'
+    },
+    'v': {
+        alias: 'version'
+    }
+};
+
+
+/**
+ * Pares the supplied arguments.
+ * The result will include arguments supplied via CLI and arguments from tsconfig.
+ */
+export function parseCommandLine(args: ReadonlyArray<string>) : ParsedCommandLine {
+    const argv = yargs
+        .usage(dedent(`tstl [options] [files...]
+
+            In addition to the options listed below you can also pass options for the typescript compiler (For a list of options use tsc -h).
+
+            NOTES:
+            - The tsc options might have no effect.
+            - Options in tsconfig.json are prioritized.`))
+        .example('tstl path/to/file.ts [...]', 'Compile files')
+        .example('tstl -p path/to/tsconfig.json', 'Compile project')
+        .options(optionDeclarations)
+        .argv;
+
+    let commandLine = ts.parseCommandLine(args);
+
+    // Run diagnostics to check for invalid tsc/tstl options
+    runDiagnostics(commandLine);
+
+    if (commandLine.options.project) {
+        findConfigFile(commandLine);
+        let configPath = commandLine.options.project;
+        let configContents = fs.readFileSync(configPath).toString();
+        const configJson = ts.parseConfigFileTextToJson(configPath, configContents);
+        commandLine = ts.parseJsonConfigFileContent(configJson.config, ts.sys, path.dirname(configPath), commandLine.options);
+    }
+
+    // Add compiler options that are ignored by TS parsers
+    for (const arg in commandLine.raw) {
+        // we dont have to check if the options is vlaid becuase we already validated at this point.
+        if (getOptionNameMap()[arg]) {
+            commandLine.options[arg] = commandLine.raw[arg];
+        }
+    }
+
+    if (commandLine.options.project && !commandLine.options.rootDir) {
+        commandLine.options.rootDir = path.dirname(commandLine.options.project);
+    }
+
+    if (!commandLine.options.rootDir) {
+        commandLine.options.rootDir = process.cwd();
+    }
+
+    if (!commandLine.options.outDir) {
+        commandLine.options.outDir = commandLine.options.rootDir;
+    }
+
+    // Run diagnostics again to check for errors in tsconfig
+    runDiagnostics(commandLine);
+
+    return commandLine;
+}
+
+interface OptionNameMap {
+    [key: string]: boolean
+}
+
+let optionNameMapCache: OptionNameMap;
+
+/**
+ * Gets a list of valid tstl options including aliases
+ * That can be used for arguemtn checking.
+ */
+function getOptionNameMap(): OptionNameMap{
+    if (optionNameMapCache) {
+        return optionNameMapCache;
+    }
+    let optNameMap: OptionNameMap = {};
+    for (let key in optionDeclarations) {
+        optNameMap[key] = true
+        let alias = optionDeclarations[key].alias;
+        if (typeof alias === "string") {
+            optNameMap[alias] = true;
+        } else {
+            alias.forEach((val) => optNameMap[val] = true);
+        }
+    }
```

Comment:
Very unclear what the purpose of this loop is and why it sets values to true, add some explanation.